cmd/nfd/subcmd/export/features.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package export
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"sigs.k8s.io/node-feature-discovery/source"
+)
+
+var (
+	outputPath string
+)
+
+func NewExportCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "features",
+		Short: "Export features for given node",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			sources := map[string]source.FeatureSource{}
+			for k, v := range source.GetAllFeatureSources() {
+				if ts, ok := v.(source.SupplementalSource); ok && ts.DisableByDefault() {
+					continue
+				}
+				sources[k] = v
+			}
+
+			// Discover all feature sources
+			for _, s := range sources {
+				if err := s.Discover(); err != nil {
+					fmt.Errorf("error during discovery of source %s: %w", s.Name(), err)
+					return err
+				}
+			}
+
+			// Make into flat format.
+			features := source.GetAllFeatures()
+			featureListing := map[string]string{}
+
+			// A flag's presence == true
+			for flagName := range features.Flags {
+				featureListing[flagName] = "true"
+			}
+
+			// Attributes are sets of elements (key value pairs)
+			for featureName, featureSet := range features.Attributes {
+				for elementName, elementValue := range featureSet.Elements {
+					featureListing[fmt.Sprintf("%s.%s", featureName, elementName)] = elementValue
+				}
+			}
+
+			for featureName, featureSet := range features.Instances {
+				for _, element := range featureSet.Elements {
+					for attrName, attrValue := range element.Attributes {
+						featureListing[fmt.Sprintf("%s.%s", featureName, attrName)] = attrValue
+					}
+				}
+			}
+			exportedLabels, err := json.MarshalIndent(featureListing, "", "    ")
```

Comment:
Currently, key collisions occur during processing. When multiple PCI devices are present on the host, only the last processed device’s attributes are retained, as all devices share the same key paths (e.g., pci.device.class). Earlier values are overwritten.

I would suggest simply executing `json.MarshalIndent(features, "", "  ")` as this:
1. eliminates key collisions
2. produces output that is already compatible with the `nfd compat` command
3. preserves the flag, attribute, and instance types which are crucial for the NFD match expression engine

However, if returning domain-style paths is required, it may be better to introduce an additional flag, such as `--output-domain` and add a counter to the path, for example: `pci.devices.1...`

Reply:
I like that interface! I will work on updating the PR for it.